The reason that the case from gh-26922 is tricky can be seen in a simpler example:
```python
In [1]: f = exp(-x**2 + x)

In [2]: f
Out[2]: 
    2    
 - x  + x
ℯ        

In [3]: integrate(f, x)
Out[3]: 
    1/4             
√π⋅ℯ   ⋅erf(x - 1/2)
────────────────────
         2          

In [4]: integrate(f, x).diff(x)
Out[4]: 
                2
 1/4  -(x - 1/2) 
ℯ   ⋅ℯ           

In [5]: integrate(f, x).diff(x).cancel()
Out[5]: 
      2
 x  -x 
ℯ ⋅ℯ  
```
The derivative expression produced from `erf` does not "look like" the original integrand. In heurisch `cancel` is used to bring it to a canonical form but that expands the exponential function resulting in terms that still do not look like the term in the integrand.

There is code for this case in heurisch:
https://github.com/sympy/sympy/blob/583993a9b99faa3f244b1863feddf4c2480ac640/sympy/integrals/heurisch.py#L431-L456
This adds the erf term but its derivative will not look like the integrand even after cancel:
https://github.com/sympy/sympy/blob/583993a9b99faa3f244b1863feddf4c2480ac640/sympy/integrals/heurisch.py#L540
The integrand itself (`f`) is not canonicalised with `cancel` when constructing the system of equations:
https://github.com/sympy/sympy/blob/583993a9b99faa3f244b1863feddf4c2480ac640/sympy/integrals/heurisch.py#L591-L593
Something like this could do it:
```diff
diff --git a/sympy/integrals/heurisch.py b/sympy/integrals/heurisch.py
index 5b9ce143bd..09e69c4053 100644
--- a/sympy/integrals/heurisch.py
+++ b/sympy/integrals/heurisch.py
@@ -588,7 +588,7 @@ def _splitter(p):
             elif isinstance(term, LambertW):
                 special[_substitute(term)] = True
 
-    F = _substitute(f)
+    F = _substitute(f.cancel())
 
     P, Q = F.as_numer_denom()
 
```
This causes a test failure but also leads to a correct result in this case rather than None:
https://github.com/sympy/sympy/blob/583993a9b99faa3f244b1863feddf4c2480ac640/sympy/integrals/tests/test_heurisch.py#L362-L365
The test failures is for this:
```
    def test_heurisch_log():
        assert heurisch(log(x), x) == x*log(x) - x
>       assert heurisch(log(3*x), x) == -x + x*log(3) + x*log(x)
E       assert None == ((-x + (x * log(3))) + (x * log(x)))
```
I'm not sure what causes that.

--- a/sympy/integrals/heurisch.py
+++ b/sympy/integrals/heurisch.py
@@ -504,8 +504,7 @@ def heurisch(f, x, rewrite=False, hints=None, mappings=None, retries=3,
         # optimizing the number of permutations of mapping              #
         assert mapping[-1][0] == x # if not, find it and correct this comment
         unnecessary_permutations = [mapping.pop(-1)]
-        # only permute types of objects and let the ordering
-        # of types take care of the order of replacement
+        # permute types of objects
         types = defaultdict(list)
         for i in mapping:
             e, _ = i
@@ -513,7 +512,8 @@ def heurisch(f, x, rewrite=False, hints=None, mappings=None, retries=3,
         mapping = [types[i] for i in types]
         def _iter_mappings():
             for i in permutations(mapping):
-                yield [j for i in i for j in i]
+                # make the expression of a given type be ordered
+                yield [j for i in i for j in ordered(i)]
         mappings = _iter_mappings()
     else:
         unnecessary_permutations = unnecessary_permutations or []

--- a/sympy/integrals/tests/test_heurisch.py
+++ b/sympy/integrals/tests/test_heurisch.py
@@ -19,7 +19,7 @@ from sympy.simplify.simplify import simplify
 from sympy.integrals.heurisch import components, heurisch, heurisch_wrapper
 from sympy.testing.pytest import XFAIL, slow
 from sympy.integrals.integrals import integrate
-from sympy import S, Matrix, And, Eq
+from sympy import S
 
 x, y, z, nu = symbols('x,y,z,nu')
 f = Function('f')
@@ -404,8 +404,9 @@ def test_heurisch_issue_26922():
 
     a, b, x = symbols("a, b, x", real=True, positive=True)
     C = symbols("C", real=True)
-    i = Integral(-C*x*exp(-a*x**2 - sqrt(b)*x), x) + Integral(C*x*exp(-a*x**2 + sqrt(b)*x), x)
-
+    i1 = -C*x*exp(-a*x**2 - sqrt(b)*x)
+    i2 = C*x*exp(-a*x**2 + sqrt(b)*x)
+    i = Integral(i1, x) + Integral(i2, x)
     res = (
         -C*exp(-a*x**2)*exp(sqrt(b)*x)/(2*a)
         + C*exp(-a*x**2)*exp(-sqrt(b)*x)/(2*a)
@@ -413,4 +414,4 @@ def test_heurisch_issue_26922():
         + sqrt(pi)*C*sqrt(b)*exp(b/(4*a))*erf(sqrt(a)*x + sqrt(b)/(2*sqrt(a)))/(4*a**(S(3)/2))
     )
 
-    assert i.doit().expand() == res
+    assert i.doit(heurisch=False).expand() == res

--- a/sympy/integrals/tests/test_integrals.py
+++ b/sympy/integrals/tests/test_integrals.py
@@ -1146,8 +1146,8 @@ def test_issue_3940():
     a, b, c, d = symbols('a:d', positive=True)
     assert integrate(exp(-x**2 + I*c*x), x) == \
         -sqrt(pi)*exp(-c**2/4)*erf(I*c/2 - x)/2
-    assert integrate(exp(a*x**2 + b*x + c), x) == \
-        sqrt(pi)*exp(c - b**2/(4*a))*erfi((2*a*x + b)/(2*sqrt(a)))/(2*sqrt(a))
+    assert integrate(exp(a*x**2 + b*x + c), x).equals(
+        sqrt(pi)*exp(c - b**2/(4*a))*erfi((2*a*x + b)/(2*sqrt(a)))/(2*sqrt(a)))
 
     from sympy.core.function import expand_mul
     from sympy.abc import k

